### PR TITLE
Use Qemu guest agent to pass cmdline info to system vms

### PR DESCRIPTION
**Pull request by remibergsma**
This replaces the socket, that requires both ends to be connected, so use the Qemu Guest Agent. The same information is passed, it works faster and more reliably now.

### CODE REVIEW

**Comment by wilderrodrigues** [comment]
go build

**Comment by wilderrodrigues** [comment]
Ping @remibergsma: could you please rebase this PR so the tests can run fine?

**Comment by remibergsma** [comment]
Thanks @wilderrodrigues rebased and pushed!

**Comment by remibergsma** [comment]
I will investigate the test failures..

**Comment by remibergsma** [comment]
Rebased with latest master. Will investigate test failures next.

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
Go build

**Comment by remibergsma** [comment]
Go build

**Comment by remibergsma** [comment]
go build

**Comment by bheuvel** [comment]
Looking into the test failures; also failed in the bubble, but also succeeded...
Gathering more info (also from here)

**Comment by bheuvel** [comment]
go build

**Comment by bheuvel** [comment]
All tests below were running on git commit: 6e31a3a44ca335aadcd9768f4205917854ca821d

Regarding tests in [Build #264](https://beta-jenkins.mcc.schubergphilis.com/job/cosmic/job/0002-cosmic-pull-request-build/264/) (mccdpod01-hv05)

```
integration.smoke.test_network.TestRebootRouter.test_reboot_router
integration.smoke.test_routers_network_ops.TestIsolatedNetworks.test_02_isolate_network_FW_PF_default_routes_egress_false
integration.smoke.test_privategw_acl.TestPrivateGwACL.test_05_rvpc_privategw_check_interface
integration.smoke.test_privategw_acl.TestPrivateGwACL.test_05_rvpc_privategw_check_interface
integration.smoke.test_ssvm.TestSSVMs.test_09_destroy_ssvm
integration.smoke.test_ssvm.TestSSVMs.test_10_destroy_cpvm
<nose.suite.ContextSuite context=TestVpcVpn>:setup
```

Regarding tests in [Build #266](https://beta-jenkins.mcc.schubergphilis.com/job/cosmic/job/0002-cosmic-pull-request-build/266/) (mccdpod01-hv07)

```
integration.smoke.test_vpc_vpn.TestVpcVpn.test_04_vpc_site2site_multiple_vpn
integration.smoke.test_vpc_vpn.TestVpcVpn.test_04_vpc_site2site_multiple_vpn
```

Regarding tests in [Build #267](https://beta-jenkins.mcc.schubergphilis.com/job/cosmic/job/0002-cosmic-pull-request-build/267/)(mccdpod01-hv05)

```
integration.smoke.test_password_server.TestPasswordService.test_02_vpc_password_service_redundant_vpc
integration.smoke.test_password_server.TestPasswordService.test_02_vpc_password_service_redundant_vpc
```

Regarding tests in [Build #270](https://beta-jenkins.mcc.schubergphilis.com/job/cosmic/job/0002-cosmic-pull-request-build/270/) (mccdpod01-hv05)

```
integration.smoke.test_privategw_acl.TestPrivateGwACL.test_05_rvpc_privategw_check_interface
integration.smoke.test_vpc_vpn.TestVpcVpn.test_03_redundant_vpc_site2site_vpn
integration.smoke.test_vpc_vpn.TestVpcVpn.test_03_redundant_vpc_site2site_vpn
integration.smoke.test_vpc_vpn.TestVpcVpn.test_04_vpc_site2site_multiple_vpn
integration.smoke.test_vpc_vpn.TestVpcVpn.test_04_vpc_site2site_multiple_vpn
```

Regarding tests in [Build #277](https://beta-jenkins.mcc.schubergphilis.com/job/cosmic/job/0002-cosmic-pull-request-build/277/) (mccdpod01-hv04)

```
integration.smoke.test_password_server.TestPasswordService.test_02_vpc_password_service_redundant_vpc
integration.smoke.test_ssvm.TestSSVMs.test_04_cpvm_internals
integration.smoke.test_ssvm.TestSSVMs.test_06_stop_cpvm
integration.smoke.test_ssvm.TestSSVMs.test_10_destroy_cpvm
```

Regarding tests in [Build #278](https://beta-jenkins.mcc.schubergphilis.com/job/cosmic/job/0002-cosmic-pull-request-build/278/) (mccdpod01-hv05)

```
integration.smoke.test_vpc_redundant.TestVPCRedundancy.test_04_rvpc_network_garbage_collector_nics
integration.smoke.test_privategw_acl.TestPrivateGwACL.test_05_rvpc_privategw_check_interface
integration.smoke.test_privategw_acl.TestPrivateGwACL.test_05_rvpc_privategw_check_interface
integration.smoke.test_ssvm.TestSSVMs.test_09_destroy_ssvm
<nose.suite.ContextSuite context=TestVpcVpn>:setup
```

Regarding tests in [Build #281](https://beta-jenkins.mcc.schubergphilis.com/job/cosmic/job/0002-cosmic-pull-request-build/281/) (mccdpod01-hv10)

```
integration.smoke.test_vpc_redundant.TestVPCRedundancy.test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL
integration.smoke.test_vpc_redundant.TestVPCRedundancy.test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL
integration.smoke.test_vpc_redundant.TestVPCRedundancy.test_02_redundant_VPC_default_routes
integration.smoke.test_vpc_router_nics.TestVPCNics.test_02_VPC_default_routes
integration.smoke.test_privategw_acl.TestPrivateGwACL.test_01_vpc_privategw_acl
```

While the master build is mostly green; out of 9 builds:
- 5 green
- 3 aborted
- 1 fail on test_06_stop_cpvm

After this PR #74 ran on hv04 (without reboot) without problems (green).
Build [Build #281](https://beta-jenkins.mcc.schubergphilis.com/job/cosmic/job/0002-cosmic-pull-request-build/281/) ran on a HV where the master build kept running, successfully...

**Comment by bheuvel** [comment]
go build

**Comment by bheuvel** [comment]
go build

**Comment by bheuvel** [comment]
Mostly excluded possible instability of the environment.
This strongly suggest the code made this instable.

Amazing how it fails, **every run** but  **on different tests**... :disappointed:

**Comment by remibergsma** [comment]
Thanks for looking into this @bheuvel. The changes made would influence whether a systemvm would boot properly or not (i.e. get the `cmdline`). I also don't get why random tests would fail.. Interesting.

**Comment by remibergsma** [comment]
@bheuvel I resolved the conflict

**Comment by remibergsma** [comment]
go build

**Comment by borisroman** [comment]
go build

**Comment by borisroman** [comment]
go build

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
By resolving the conflict I probably broke something, as the build hangs on 'template 'tiny linux kvm' is not Ready'. Will investigate on Monday.

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
@bheuvel Spent some time on this PR again. I think the link between the random test failures and the above issue (ssvm didn't start so build was hanging) is that for some reason a systemvm sometimes isn't starting properly. When this happens, Cosmic cannot SSH and eventually then the test failed. The systemvm was booted but without config (hostname: systemvm). Needs a bit more investigation obviously but we don't need to look at the tests for now I'd say. To be continued..

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
go build

**Comment by remibergsma** [comment]
Squashed and rebased PR

**Comment by remibergsma** [comment]
It's green now! Eventually changed it to put the authorized_keys file and the cmdline into their own files and sent another to mark when the process is done. The files are read back to verify that they've been written ok. Also, the script exits as soon as something goes wrong. The mgt server will then retry anyway and it prevents a halfway patched system VM.

**Comment by bheuvel** [comment]
go build

**Comment by bheuvel** [comment]
Cool! Great work!
Considering the unstable results of before, (re) started a build to show it's really stable now.

File: cosmic-core/systemvm/patches/debian/config/etc/init.d/cloud-early-config [hunk 1]
```diff
@@ -99,40 +99,57 @@ get_boot_params() {
...
-            then
-              systemvm_type=$(cat /var/cache/cloud/cmdline | tr ' ' '\n' | grep type | cut -d\= -f2)
-              if [[ "${systemvm_type}" == "secstorage" || "${systemvm_type}" == "consoleproxy" ]];
-              then
-                log_it "Found a non empty cmdline file for type = ${systemvm_type}. Will now exit the loop and proceed with configuration."
-                break;
-              fi
+        # Make sure the Qemu guest agent is running, or else we will be waiting forever
+        ps aux | grep qemu | grep -v grep > /dev/null 2>&1
+        if [ $? -gt 0 ]; then
```
Comment (sspans) [inline cosmic-core/systemvm/patches/debian/config/etc/init.d/cloud-early-config:110]: These can be replaced by:
`
if ! pgrep /usr/sbin/qemu-ga &>/dev/null; then
`

Or even better:

`
if ! /etc/init.d/qemu-guest-agent status &>/dev/null; then
`

Comment (remibergsma) [reply cosmic-core/systemvm/patches/debian/config/etc/init.d/cloud-early-config:110]: I will address this in a separate PR so we can proceed with this now

**Pull request merged**